gateway/src/apicast/policy/soap/policy.lua

```diff
@@ -0,0 +1,68 @@
+--- SOAP Policy
+-- This policy adds support for a very small subset of SOAP.
+-- This policy basically expects a SOAPAction URI in the SOAPAction header or
+-- the content-type header.
+-- The SOAPAction header is used in v1.1:
+-- https://www.w3.org/TR/2000/NOTE-SOAP-20000508/#_Toc478383528, whereas the
+-- Content-Type header is used in v1.2:
+-- https://www.w3.org/TR/soap12-part2/#ActionFeature
+-- The SOAPAction URI is matched against the mapping rules defined for the
+-- service and the usage resulting from that is authorized and reported
+-- against 3scale's backend.
+
+local sub = string.sub
+local len = string.len
+
+local policy = require('apicast.policy')
+local _M = policy.new('SOAP policy')
+
+local soap_action_header = 'SOAPAction'
+local soap_action_ctype = 'application/soap+xml?action='
+
+local new = _M.new
+
+local function starts_with(str, start)
+  return sub(str, 1, len(start)) == start
+end
+
+local function soap_action_in_header(headers)
+  return headers[soap_action_header]
+end
+
+local function soap_action_in_ctype(headers)
+  local ctype = headers['Content-Type']
+
+  if ctype and starts_with(ctype, soap_action_ctype) then
+    return sub(ctype, len(soap_action_ctype) + 1, -1)
+  else
+    return nil
+  end
+end
+
+local function extract_soap_uri()
+  local headers = ngx.req.get_headers() or {}
+  return soap_action_in_ctype(headers) or soap_action_in_header(headers)
+end
+
+--- Initialize a SOAP policy
+-- @tparam[opt] table config
+function _M.new(config)
+  return new(config)
+end
+
+--- Rewrite phase
+-- Adds to the field 'add_to_usage' of the shared context a table that
+-- contains pairs of (metric, usage).
+-- @tparam table context Shared context between policies
+function _M.rewrite(_, context)
+  local soap_action_uri = extract_soap_uri()
+
+  if soap_action_uri then
+    local soap_usage = context.service:get_usage(
+      ngx.req.get_method(), soap_action_uri)
```

Comment:
So this is using service mapping rules? I'm not sure users can define full URLs there.
I though we will have some mapping rules in this policy.

Reply:
I assumed it was possible to define something like `/a_path#some_action` in the mapping rules. That's why I decided to do it this way. I see now that it's not possible. With that constraint, I think we need to include mapping rules in the configuration of this policy.

Reply:
@mikz  Let me know what you think about this solution:
- Add a mapping rules array in the policy config. Each element will contain: `pattern`, `http_method`, `metric_name`, and `delta`. The same fields we have for the service mapping rules.
- Add a `add_mapping_rule()` method in `Service`. This would simply accept the fields above, add the ones needed to construct the rule, and then add it to the `self.rules` table. We'll need to make sure that `add_mapping_rule()` is only called once during the lifetime of the policy so we don't add duplicated rules.
- The above will let us call the existing `Service:get_usage()`. This method returns the metrics to auth and report based on the mapping rules defined. We'll call this method only when we receive a SOAP action.

Reply:
@davidor interesting idea. I'm not sure if I like the idea of merging then into one list. I think they should be applied separately. IMO there could be some rules that match both and because some are absolute and some relative it could be quite nasty as `/` would increment hits all the time and twice.
Maybe we need to expose`UsageExtractor:get_usage` which is initialized with list of mapping rules. Basically exposing the low level API where you can inject own mapping rules.

Reply:
I get your first point, but wouldn't that be coherent with the way mapping rules work currently?

Suppose that the policy adds this mapping rule: `/a_path#a_soap_action => +1 a_soap`. If the user has already defined `/a_path => +1 a_path`, we should increase both `a_soap` and `a_path` by 1. If the user already defined `/ => +1 hits`, and `hits` is a parent metric of the other 2, well, it's going to be increased by 3(!), but that's how it currently works no?

Regarding your second suggestion, could you please expand a bit on it or provide some pseudo-code or examples?

Reply:
I think the issue is when you have following rules:

/ => hits
/a => a

And in soap you'll have 
http://example.com/soap#a => a

If you match `http://example.com/soap#a` it will return `hits=1&a=1`.

So it depends if this **replaces** mapping rules or **complements** them.
If it would first match mapping rules and then soap rules and merged the results then for a request:
```http
GET /
Content-Type: application/xml+soap?action=http://example.com/soap#a
```

it would match `hits=2&a=1`. Which is wrong IMO. Because the path mapping rules should not be applied to soap mapping rules (because they are relative and not absolute).
I'm not thinking about parent/child metrics, just about applying mapping rules to the request path/ soap action.
The point is that current path relative mapping rules like `/a` can unintentionally match soap services like `http://example.com/a` and double counting. And also match different services like `http://example.com/a` and `http://foobar.com/a`.

And the signature of UsageExtractor would be quite simple. 
```
local usage_rules = UsageExtractor.new({
  { pattern = 'http://example.com/a', http_method = 'GET', metric_name = 'a', delta = 1 }
})
local usage = usage_rules:get_usage(...)
```

And the `Service:get_usage` would use this too.  But we would just have an API to create own independent list of rules.

Reply:
@mikz  Thanks for the example.

It's clear to me now why you think that a url received via SOAP Action should not be matched against the service mapping rules. It should only be matched against the mapping rules specified in the SOAP policy. I agree with that.

Now, what happens when the SOAP policy is enabled but a request does not contain a SOAP Action ? I think we should apply the service mapping rules in that case.

I like the idea of extracting a `UsageExtractor` module from `Service` and `Configuration`. This is something I've been thinking about. Those 2 modules have many responsabilities and need to be splitted. This seems to be a good excuse to do it :) I'll address this task in a separate PR.